ddtrace/internal/codeowners.py

```diff
@@ -0,0 +1,191 @@
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+    """
+    regex = ""
+
+    slash_pos = pattern.find("/")
+    anchored = slash_pos > -1 and slash_pos != len(pattern) - 1
+
+    regex += r"\A" if anchored else r"(?:\A|/)"
+
+    matches_dir = pattern[-1] == "/"
+    pattern_trimmed = pattern.strip("/")
+
+    in_char_class = False
+    escaped = False
+
+    iterator = enumerate(pattern_trimmed)
+    for i, ch in iterator:
+
+        if escaped:
+            regex += re.escape(ch)
+            escaped = False
+            continue
+
+        if ch == "\\":
+            escaped = True
+        elif ch == "*":
+            if i + 1 < len(pattern_trimmed) and pattern_trimmed[i + 1] == "*":
+                left_anchored = i == 0
+                leading_slash = i > 0 and pattern_trimmed[i - 1] == "/"
+                right_anchored = i + 2 == len(pattern_trimmed)
+                trailing_slash = i + 2 < len(pattern_trimmed) and pattern_trimmed[i + 2] == "/"
+
+                if (left_anchored or leading_slash) and (right_anchored or trailing_slash):
+                    regex += ".*"
+
+                    next(iterator, None)
+                    next(iterator, None)
+                    continue
+            regex += "[^/]*"
+        elif ch == "?":
+            regex += "[^/]"
+        elif ch == "[":
+            in_char_class = True
+            regex += ch
+        elif ch == "]":
+            if in_char_class:
+                regex += ch
+                in_char_class = False
+            else:
+                regex += re.escape(ch)
+        else:
+            regex += re.escape(ch)
+
+    if in_char_class:
+        raise ValueError("unterminated character class in pattern {pattern}".format(pattern=pattern))
+
+    regex += "/" if matches_dir else r"(?:\Z|/)"
+    return re.compile(regex)
+
+
+class Codeowners(object):
+    """Provide interface to parse CODEOWNERS file and match a given path against it."""
+
+    KNOWN_LOCATIONS = (
+        "CODEOWNERS",
+        ".github/CODEOWNERS",
+        "docs/CODEOWNERS",
+        ".gitlab/CODEOWNERS",
+    )
+
+    def __init__(self, path=None, cwd=None):
+        # type: (Optional[str], Optional[str]) -> None
+        """Initialize Codeowners object.
+
+        :param path: path to CODEOWNERS file otherwise try to use any from known locations
+        """
+        path = path or self.location(cwd)
+        if path is not None:
+            self.path = path  # type: str
+        self.patterns = []  # type: List[Tuple[re.Pattern, List[str]]]
+        self.parse()
+
+    def location(self, cwd=None):
+        # type: (Optional[str]) -> Optional[str]
+        """Return the location of the CODEOWNERS file."""
+        cwd = cwd or os.getcwd()
+        for location in self.KNOWN_LOCATIONS:
+            path = os.path.join(cwd, location)
+            if os.path.exists(path):
```

Comment:
```suggestion
            if os.path.isfile(path):
```

? 🤷🏻 either is probably fine. this suggestion is not required, feel free to dismiss